http-clients/apache-async-client/src/main/java/software/amazon/awssdk/http/apache/async/ApacheAsyncRequestFactory.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.apache.async;
+
+import java.net.URI;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import org.apache.hc.core5.http.ContentType;
+import org.apache.hc.core5.http.HttpHeaders;
+import org.apache.hc.core5.http.nio.AsyncEntityProducer;
+import org.apache.hc.core5.http.nio.AsyncRequestProducer;
+import org.apache.hc.core5.http.nio.support.AsyncRequestBuilder;
+import org.apache.hc.core5.reactive.ReactiveEntityProducer;
+import software.amazon.awssdk.annotations.ReviewBeforeRelease;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.awssdk.http.async.AsyncExecuteRequest;
+import software.amazon.awssdk.utils.http.SdkHttpUtils;
+
+/**
+ * Responsible for creating Apache HttpClient 5 request objects.
+ */
+@SdkInternalApi
+class ApacheAsyncRequestFactory {
+    private static final List<String> IGNORE_HEADERS = Collections.singletonList(HttpHeaders.HOST);
+
+    public AsyncRequestProducer create(AsyncExecuteRequest request, boolean isHttp2) {
+        URI uri = request.request().getUri();
+        AsyncRequestBuilder base = createApacheRequest(request, uri.toString());
+        addHeadersToRequest(base, request.request(), isHttp2);
+
+        return base.build();
+    }
+
+    private AsyncRequestBuilder createApacheRequest(AsyncExecuteRequest request, String uri) {
+        switch (request.request().method()) {
+            case HEAD:
+                return AsyncRequestBuilder.head(uri);
+            case GET:
+                return AsyncRequestBuilder.get(uri);
+            case DELETE:
+                return AsyncRequestBuilder.delete(uri);
+            case OPTIONS:
+                return AsyncRequestBuilder.options(uri);
+            case PATCH:
+                return AsyncRequestBuilder.patch(uri).setEntity(wrapEntity(request));
+            case POST:
+                return AsyncRequestBuilder.post(uri).setEntity(wrapEntity(request));
+            case PUT:
+                return AsyncRequestBuilder.put(uri).setEntity(wrapEntity(request));
+            default:
+                throw new RuntimeException("Unknown HTTP method name: " + request.request().method());
```

Comment:
Why not IllegalArgument or IllegalState?

Reply:
I inherited this choice of exception from `ApacheHttpRequestFactory`, part of the `apache-client` (synchronous) integration that I forked. I'm generally not too picky about exception types, so I didn't change it.